instance/models/server.py

```diff
@@ -242,18 +252,36 @@ def _set_progress(self, progress, expected_status=None):
         # The '_progress' field needs to be saved:
         self.save()
 
-    def sleep_until_status(self, *target_status_list):
+    def sleep_until(self, condition):
         """
-        Sleep in a loop until the server reaches one of the specified statuses
+        Sleep in a loop until condition related to server status is fulfilled.
+
+        Raises an exception if the desired condition can not be fulfilled.
+        This can happen if the server is in a steady state (i.e., a state that is not expected to change)
+        that does not fulfill the desired condition.
+
+        Use as follows:
+
+            server.sleep_until(lambda: server.status.is_steady_state)
+            server.sleep_until(lambda: server.status.accepts_ssh_commands)
         """
-        self.logger.info('Waiting to reach status %s...', target_status_list)
+        if condition():
+            return
+        else:
+            if self.status.is_steady_state:
+                raise SteadyStateException(
+                    "The current state ({name}) does not fulfill the desired condition "
+                    "and is steady, i.e., it is not expected to change.".format(name=self.status.name)
+                )
```

Comment:
This check should be inside the `while` loop. When `self.update_status()` is called, the server could transition into a steady state, and then this would loop forever.